hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/telemetry/TraceExtractingMessageInterceptor.kt

```diff
@@ -0,0 +1,70 @@
+package uk.gov.justice.hmpps.sqs.telemetry
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.awspring.cloud.sqs.MessageHeaderUtils
+import io.awspring.cloud.sqs.listener.interceptor.MessageInterceptor
+import io.opentelemetry.api.GlobalOpenTelemetry
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.api.trace.SpanKind
+import io.opentelemetry.context.Context
+import io.opentelemetry.context.Scope
+import io.opentelemetry.context.propagation.TextMapGetter
+import org.springframework.messaging.Message
+
+/**
+ * Intercepts messages before they are passed to the `@SqsListener`.
+ *
+ * On receiving a message, this interceptor will:
+ *  1. attempt to extract the trace context from the message attributes,
+ *  2. start a new OpenTelemetry span with the name "RECEIVE $eventType",
+ *  3. end the span once the message has finished processing.
+ */
+class TraceExtractingMessageInterceptor(private val objectMapper: ObjectMapper) : MessageInterceptor<Any> {
+  override fun intercept(message: Message<Any>): Message<Any> {
+    val payload = message.payload
+    if (payload !is String) return message
+
+    val attributes = objectMapper.readValue(
+      objectMapper.readTree(payload).at("/MessageAttributes").traverse(),
+      object : TypeReference<MutableMap<String, MessageAttribute>>() {},
+    )
+    val spanName = attributes["eventType"]?.let { "RECEIVE ${it.Value}" } ?: "RECEIVE"
+    val span = attributes.extractTelemetryContext().startSpan(spanName)
```

Comment:
We've been using the @WithSpan annotation on our listeners to get message receive into App Insights... [like this](https://github.com/ministryofjustice/hmpps-prisoner-from-nomis-migration/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/activities/ActivitiesMigrationMessageListener.kt#L29).

I guess we'll be able to drop these now?

Reply:
Yes you can drop these now, but no harm leaving them in either - they'll display as a child span of the "RECEIVE".